### ISSUE TITLE
node_access filters out accessible nodes when node is left joined

### ISSUE THREAD

**Issue by aitala** [backdrop/backdrop-issues#6421]
## Description of the bug

From https://www.drupal.org/project/drupal/issues/1349080 and https://www.drupal.org/node/3176634

Currently, when a node table is left joined into a query, adding a node_access tag to the query filters out accessible rows from the base table (effectively acting more like an inner join). In particular, rows containing null values are incorrectly filtered out by node_access (i.e., if the base table has null entries for the node ID, node_access denies access to all users based on those null values, even though they should have no relevance to node_access checks). The most common example is a view where rows disappear from the table after adding an optional relationship to the view.

The patch in https://www.drupal.org/files/issues/2023-06-12/drupal-node_access_for_left_joins_d7.3176634-34.patch appears to work after stripping out the testing changes

## Steps To Reproduce

1. enable Views
2. Create a field referencing nodes
3. Create a node having the entity reference field empty,
4. Create a View with a relationship using the entity reference field. Do NOT select "Require this relationship".
5. Expected result is that regular users can see the node in the View. Current result is that regular users cannot see the node.

## Additional information

Add any other information that could help, such as:
- Backdrop CMS version: 1.27.1

**Comment by laryn** [comment]
This was [recently committed for Drupal 10.3](https://git.drupalcode.org/project/drupal/-/commit/c271adbb22f28a66ff4321b2fb7f86f20fe0c838). The patch above is the backport and should apply pretty cleanly. I haven't looked at all the tests that are included in the patch, but nice to have them ready to go, too!

**Comment by laryn** [comment]
@aitala I just ported the patch here, including tests, so let's see how it looks after tests have been run on the PR.

**Comment by argiepiano** [comment]
I've been doing some testing and digging. The patch in D7 actually works as expected, but not in Backdrop.

I have traced the problem. There was a fix 2 years ago in D7 that did not make it into Backdrop:
https://www.drupal.org/project/drupal/issues/2582069

That PR fixed the the fact that conditions within joins were not compiled when the SelectQuery was compiled before execution.

This means that, while the patch here is correct, it will actually throw a fatal error since the compile query is faulty (as he node_access condition added in the leftJoin is not compiled correctly). The patch here will only works if that is fixed.

So, we need to open another issue to fix that problem, and once that's merged, this will be unblocked.

**Comment by argiepiano** [comment]
One more piece of info: that D7 patch that fixed the compilation of conditions with joins was mentioned in a list of crossports needed to Backdrop that actually was never completed:

https://github.com/backdrop/backdrop-issues/issues/5770

**Comment by argiepiano** [comment]
I have created an issue to fix the bug I mentioned about, which is blocking this PR.

https://github.com/backdrop/backdrop-issues/issues/6852

**Comment by laryn** [comment]
@kswan I know you were going to try to take a look at this, so I've rebased and pushed up what I had on my local. The only difference with the tests from the Drupal 7 patch this started from is that it doesn't try to fudge a Drupal 10 view -- I created a view and some entity reference fields and added them to the config folder in the test module that gets enabled. The idea is they will get installed when the module is enabled, and we can use the actual results shown by the view instead of fudging it. I don't know how close it is to being functional, but it seems a cleaner way to go than what they ended up doing in the D7 patch. I was also planning to use the core "view own unpublished content" as the permission to mark something as public/private depending who was viewing it. Does that sound like something you can start with? 

If you prefer to start from scratch, there's no requirement you start from my PR of course. I just added this in case it is useful.

**Comment by argiepiano** [comment]
I have tested this updated/rebased patch, and, after fixing core with https://github.com/backdrop/backdrop-issues/issues/6852, it solves the issue of the disappearing, left joined, nodes in the View. WFM!

**Comment by argiepiano** [comment]
Sorry, I just tried the test that was added in the PR and it doesn't work. The functionality does. Marking NW again so that we can fix the test.

**Comment by argiepiano** [comment]
@laryn, I've been playing with the PR trying to fix the new test (the new functionality actually works great). I'm facing some issues with your approach:
- The entity reference fields cannot be created from config
- They can only be created in hook_install(), AND only if you invoke `  _field_info_collate_types_reset();` to reset field type caches before creating them. There is an issue with Backdrop that cannot find the `entityreference` field type unless you reset the caches
- After solving that, I'm trying for the view to work - no matter what I do, it's not showing the list of nodes - it's always empty. I think there is something wrong with the View (although it seems to work outside the test)

Wondering if you really want to stick with this way of testing? There may be a simpler, more direct way by using a `LEFT JOIN` in a database query, instead of the heavier, less reliable way of creating an actual view.

**Comment by laryn** [comment]
I am not wedded to this approach. I think it’s close to what was done in D10 but as long as we test what we need to test, I’m okay.

**Comment by argiepiano** [comment]
@laryn, I've played around quite a bit - the view approach was causing quite a bit of problems. I decided to try the D7 version of the test, which mimics a view but provides a page output themed as a table, with a `header` attribute that mimics the D10 view output.

In my local testing, this is now working great - the new tests passes with no problems. Of course, here it will fail, because we need to merge   https://github.com/backdrop/backdrop-issues/issues/6852 first in order for this fix (and new test) to work.

We will not know if this is solid until that is merged.

**Comment by argiepiano** [comment]
I'm seeing test failures for `BookTestCase` and `SearchNodeAccessTest`, but those are passing locally for me. My guess is that those test failures will disappear once #6852 is merged.

**Comment by argiepiano** [comment]
@laryn thanks for rebasing. Tests are now passing 🎉 

I can fix the style issues when I get home later today.

**Comment by kswan** [comment]
I tested https://github.com/backdrop/backdrop/pull/5018 and it works perfectly for me. 

When the referenced node field is blank, the view correctly shows the parent node with the referenced node field blank.
When the person doesn't have access to the referenced node, the parent node is shown, and the restricted referenced node field is blank. 

Adding WFM.

**Comment by kswan** [comment]
I closed the old PRs. It was a little confusing with three open PRs. 

Thank you, @argiepiano!

**Comment by laryn** [comment]
The code looks good to me. I made one minor comment/suggestion in a docblock in the tests @argiepiano.

**Comment by quicksketch** [comment]
Thank you @argiepiano, @laryn, @kswan, and @aitala! Also thanks to @izmeez for the work on the Drupal issue [#3176634](https://www.drupal.org/node/3176634), on which this is based! Although that issue may never be completed for Drupal 7, it's always nice to see work not gone to waste and gain new life in our project.

This might have the longest commit credit I've ever done: https://github.com/backdrop/backdrop/commit/fc29dab549fb6bdb802316675a79ce456c7ff4ec 

> Issue #6421: Prevent node_access from hiding left-joined nodes.
> 
> By @argiepiano, @kswan, @laryn, @izmeez, and @aitala.
> 
> Port of Drupal issue #3176634 by joseph.olstad, poker10, laryn, brad.bulger, mustanggb, mustanggb, voleger, izmeez, amir simantov, joelpittet.
> 
> Port of Drupal #1349080 by Nephele, freelock, Spokje, quietone, Lendude, agentrickard, tic2000, jofitz, RunePhilosof, manuelBS, AndreyMaximov, aleevas, dmsmidt, ChristianAdamski, ducktape, lokapujya, kalistos, ekes, maximpodorov, jordan.jamous, mgifford, fuerst, daffie, markhalliwell, emorency, thetwentyseven, damiankloip, mohit1604, rensingh99, gilsbert, xjm, andypost, tim.plunkett, Renee S, TwoD, rbruhn, mxr576, MustangGB, alexpott, dawehner, smustgrave, catch, lauriii, larowlan, merlinofchaos, Alan D., TBarina, jhedstrom, neclimdul, pookmish, danflanagan8, Les Lim, cherabomb, Christopher Riley, chx.

Merged into 1.x and 1.31.x. Thank you thank you everyone!
